### ISSUE TITLE
data module

### ISSUE THREAD

**Issue by Choumiko** [Afforess/Factorio-Stdlib#9]
A module to allow easier data.raw manipulation:
```lua
 -- creates a new pole named my-op-pole based on big pole, changing mining result name as well
local op_pole = Data.copy("electric-pole", "big-electric-pole", "my-op-pole")
op_pole.wire_distance = 9001

local recipe = Data.copy("recipe", "big.electric-pole", "my-op-pole")
local item = Data.copy("item", "big.electric-pole", "my-op-pole")
-- could perform some checks, e.g. duplicate name
Data.extend{op_pole, recipe, item}
Data.add_unlock_to_tech("some-tech", recipe) -- or Data.add_unlock_to_tech("some-tech", "my-op-pole")

-- remove all 'references' to the item from data.raw
Data.remove("item", "name-to-remove")
--replace all 'references' to itemA with itemB
Data.replace("item", "itemA", "itemB")
```
Last 2 seem like edge cases

**Comment by Afforess** [comment]
Yes, definitely. I've written a few things like this before for Marathon. Data manipulation is often a pain, and it would be nice to have easy functions for.

**Comment by Afforess** [comment]
FYI, I recently started on this. I went with a jquery selector ... inspired solution. Here's the commit: https://github.com/Afforess/Factorio-Stdlib/commit/bca9d7886b28137995133d90ba97a840d1ce143a

Example:
`Data.select("recipe:copper.*) -- selects all recipes matching copper.*`
`Recipe.select("copper.*) -- selects all recipes matching copper.*`
`Recipe.select("copper.*:ingredients:copper.*") -- selects all ingredients whose name matches 'copper.*' in recipes matching copper.*`

**Comment by Afforess** [comment]
I also wrote a metatable that is dynamically assigned to the recipe fields 'ingredients' and 'results', which detects the layout of the item, and emulates the 'amount' and 'name' fields when the item object is an array, or emulates the '1' and '2' fields if the item is a table. This should allow modders to stop trying to write code for either way and detecting, and just use the named fields when accessing, but write the shorthand when updating recipes.

**Comment by Afforess** [comment]
The very basics of this was added to 0.6.0, but more work needs to be done to support specific data structures.

https://github.com/Afforess/Factorio-Stdlib/releases/tag/0.6.0

**Comment by boweiliu** [comment]
What's our progress so far? What remains to be done on this issue?
